### PR TITLE
parametric: added parametric tests for otel api

### PR DESCRIPTION
**Pull request by lievan**
## Description

Added tests for starting/ending parent and child spans, force flushing, and setting span status

## Check list

Your PR is not ready to be reviewed? Please save it as a draft :pray:

- [ ] Follows the style guidelines of this project (See [how to easily lint the code](https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md))
- [ ] CI is passing
- [ ] There is at least one approval from code owners

Yes to all? Feel free to merge it whenever you want :heart:

### CODE REVIEW

File: parametric/apps/golang/otel_methods.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	}
+	if args.GetNewRoot() {
+		otelOpts = append(otelOpts, ot_api.WithNewRoot())
+	}
+	if t := args.GetTimestamp(); t != 0 {
+		tm := time.Unix(t, 0)
+		otelOpts = append(otelOpts, ot_api.WithTimestamp(tm))
+	}
+
+	// if attrs := args.GetAttributes(); attrs != nil {
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:36]: Current api allows passing in the attributes at the start, it's fin to continue with it

File: parametric/apps/golang/otel_methods.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		}
+		// all values are represented as slices
+		first := lv.GetVal()[0]
+		switch first.Val.(type) {
+		case *AttrVal_StringVal:
+			inp := make([]string, n)
+			for _, v := range lv.GetVal() {
+				inp = append(inp, v.GetStringVal())
+			}
+			span.SetAttributes(attribute.StringSlice(k, inp))
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:78]: The most common scenario would be a singular bool, int, etc.
```suggestion
			inp := make([]string, n)
			for _, v := range lv.GetVal() {
				inp = append(inp, v.GetStringVal())
			}
			if len(inp) == 1{
			span.SetAttributes(attribute.String(k, inp[0]))
			}
			span.SetAttributes(attribute.StringSlice(k, inp))
```

File: parametric/apps/golang/otel_methods.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+				inp = append(inp, v.GetIntegerVal())
+			}
+			span.SetAttributes(attribute.Int64Slice(k, inp))
+		}
+
+	}
+	return &OtelSetAttributesReturn{}, nil
+}
+
+func (s *apmClientServer) OtelSetName(ctx context.Context, args *OtelSetNameArgs) (*OtelSetNameReturn, error) {
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:103]: A note to check - Otel component == ddtrcae operation name

File: parametric/apps/golang/otel_methods.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		fmt.Sprintf("EndOtelSpan call failed, span with id=%s not found", args.SpanId)
+	}
+	span.SetName(args.Name)
+	return &OtelSetNameReturn{}, nil
+}
+
+func (s *apmClientServer) OtelFlushSpans(ctx context.Context, args *OtelFlushSpansArgs) (*OtelFlushSpansReturn, error) {
+	s.otelSpans = make(map[string]ot_api.Span)
+	success := false
+	set_flush_success := func(ok bool) { success = ok }
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:115]: if it's used once, we don't really need it

File: parametric/apps/golang/otel_methods.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	s.otelSpans = make(map[string]ot_api.Span)
+	return &OtelFlushTraceStatsReturn{}, nil
+}
+
+func (s *apmClientServer) OtelIsRecording(ctx context.Context, args *OtelIsRecordingArgs) (*OtelIsRecordingReturn, error) {
+	span, ok := s.otelSpans[args.SpanId]
+	if !ok {
+		fmt.Printf("IsRecording call failed, span with id=%s not found", args.SpanId)
+	}
+	is_recording := span.IsRecording()
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:130]: Same here with one-use variable

File: parametric/apps/golang/otel_methods.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	return &OtelIsRecordingReturn{IsRecording: is_recording}, nil
+}
+
+func (s *apmClientServer) OtelSpanContext(ctx context.Context, args *OtelSpanContextArgs) (*OtelSpanContextReturn, error) {
+	span, ok := s.otelSpans[args.SpanId]
+	if !ok {
+		fmt.Printf("SpanContext call failed, span with id=%s not found", args.SpanId)
+	}
+	span_context := span.SpanContext()
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:139]: Same here with one-use variable 😁

File: parametric/apps/golang/otel_methods.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+		span.SetStatus(codes.Error, args.Description)
+	} else if args.Code == "OK" {
+		span.SetStatus(codes.Ok, args.Description)
+	} else {
+		fmt.Sprintf("Invalid code")
+	}
+	return &OtelSetStatusReturn{}, nil
+}
+
+//func (s *apmClientServer) StopOtelTracer(context.Context, *StopOtelTracerArgs) (*StopOtelTracerReturn, error) {
```
Comment (dianashevchenko) [inline parametric/apps/golang/otel_methods.go:172]: Can probably be removed

File: parametric/apps/golang/tracer_methods.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+func (s *apmClientServer) StartSpan(ctx context.Context, args *StartSpanArgs) (*StartSpanReturn, error) {
+	var opts []tracer.StartSpanOption
+	if args.ParentId != nil && *args.ParentId > 0 {
```
Comment (dianashevchenko) [inline parametric/apps/golang/tracer_methods.go:10]:
```suggestion
if args.GetParentID() > 0{
```

File: parametric/apps/golang/tracer_methods.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		parent := s.spans[*args.ParentId]
+		opts = append(opts, tracer.ChildOf(parent.Context()))
+	}
```
Comment (dianashevchenko) [inline parametric/apps/golang/tracer_methods.go:13]:
```suggestion
if args.GetParentID() > 0{
```

File: parametric/test_otel_tracer.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
+import time
+import pytest
+
+from parametric.spec.trace import find_trace_by_root
+from parametric.spec.trace import find_span_in_traces
+from parametric.spec.trace import find_span
+from parametric.spec.otel_trace import OtelSpan
+
+# todo: add prefix_library to run otel_*
```
Comment (dianashevchenko) [inline parametric/test_otel_tracer.py:9]: better to move constants to the spec file

**Review by dianashevchenko — CHANGES_REQUESTED** [review]

**Review by dianashevchenko — CHANGES_REQUESTED** [review]

File: parametric/test_otel_tracer.py [hunk 11]
```diff
@@ -1,13 +1,186 @@
...
+from parametric.spec.trace import find_trace_by_root
+from parametric.spec.trace import find_span_in_traces
+from parametric.spec.trace import find_span
+from parametric.spec.otel_trace import OtelSpan
+from parametric.spec.otel_trace import OTEL_UNSET_CODE, OTEL_ERROR_CODE
+from parametric.spec.otel_trace import SK_PRODUCER
 
-def test_start_otel_api_span_with_set_name(test_agent, test_library):
-    """Test span is created with OTel API of Datadog Tracing Library"""
+# todo: add prefix_library to run otel_*
```
Comment (dianashevchenko) [inline parametric/test_otel_tracer.py:11]:
```suggestion
```

File: parametric/test_otel_tracer.py [hunk 12]
```diff
@@ -1,13 +1,186 @@
...
+    trace = find_trace_by_root(traces, OtelSpan(name="operation"))
+    assert len(trace) == 1
+
+    root_span = find_span(trace, OtelSpan(name="operation"))
+    assert root_span["meta"]["env"] == "otel_env"
+    assert root_span["service"] == "otel_serv"
+    assert root_span["name"] == "operation"
+    assert root_span["resource"] == "operation"
+    assert root_span["meta"]["start_attr_key"] == "start_attr_val"
+    assert root_span["duration"] * (1e-9) == end_time - start_time
```
[the remark this answers is not in the document]
Comment (lievan) [reply parametric/test_otel_tracer.py]: Created a separate file `test_otel_span_methods.py` to increase granularity!

File: parametric/test_otel_span_methods.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            "operation", span_kind=SK_PRODUCER, timestamp=start_time, new_root=True, attributes=starting_attributes,
+        ) as parent:
+            parent.finish(timestamp=start_time + duration_s)
+    duration_ns = duration_s / (1e-9)
+
+    traces = test_agent.wait_for_num_traces(1)
+    trace = find_trace_by_root(traces, OtelSpan(name="operation"))
+    assert len(trace) == 1
+
+    root_span = find_span(trace, OtelSpan(name="operation"))
```
Comment (dianashevchenko) [inline parametric/test_otel_span_methods.py:41]: span = get_span(test_agent)

File: parametric/test_otel_span_methods.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+    # entering test_otel_library starts the tracer with the above options
+    with test_library:
+        duration_s = int(2 * 1000000)
+        start_time = int(time.time())
+        starting_attributes = {"start_attr_key": "start_attr_val"}
+        with test_library.start_otel_span(
+            "operation", span_kind=SK_PRODUCER, timestamp=start_time, new_root=True, attributes=starting_attributes,
+        ) as parent:
+            parent.finish(timestamp=start_time + duration_s)
```
Comment (dianashevchenko) [inline parametric/test_otel_span_methods.py:34]: is it really a parent?

**Comment by lievan** [comment]
Worked moved to https://github.com/DataDog/system-tests/pull/884

**Pull request closed without merging**